lib/srv/server/ec2_watcher.go

```diff
@@ -374,36 +415,38 @@ func newEC2InstanceFetcher(cfg ec2FetcherConfig) *ec2InstanceFetcher {
 		safeUpdateGroup := shsprintf.EscapeDefaultContext(cfg.UpdateGroup)
 		envVars = append(envVars, "TELEPORT_UPDATE_GROUP="+safeUpdateGroup)
 	}
-	if len(envVars) > 0 {
-		parameters["env"] = strings.Join(envVars, " ")
+
+	// For custom SSM Documents, the following parameters are used: env, scriptName and token.
+	//
+	// However, when using the pre-defined SSM Document AWS-RunShellScript, we need to construct
+	// the "commands" parameter.
+	if cfg.Document != types.AWSSSMDocumentRunShellScript {
+		return ssmRunCommandParametersForCustomDocuments(cfg, envVars)
 	}
 
-	fetcher := ec2InstanceFetcher{
-		EC2ClientGetter:     cfg.EC2ClientGetter,
-		Filters:             tagFilters,
-		Region:              cfg.Region,
-		DocumentName:        cfg.Document,
-		Parameters:          parameters,
-		Integration:         cfg.Integration,
-		DiscoveryConfigName: cfg.DiscoveryConfigName,
-		EnrollMode:          cfg.EnrollMode,
-		cachedInstances: &instancesCache{
-			instances: map[cachedInstanceKey]struct{}{},
-		},
+	publicProxyAddr := cmp.Or(cfg.Matcher.Params.PublicProxyAddr, cfg.ProxyPublicAddr)
+
+	command := fmt.Sprintf("curl -s -L https://%s/v1/webapi/scripts/installer/%s | bash -s %s",
+		publicProxyAddr,
+		cfg.Matcher.Params.ScriptName,
+		cfg.Matcher.Params.JoinToken,
+	)
```

Comment:
I don't have the full context here, so just double checking: Should this also use `shsprintf.EscapeDefaultContext` like `safeUpdateGroup` above?

```go
  safeProxyAddr := shsprintf.EscapeDefaultContext(publicProxyAddr)
  safeScriptName := shsprintf.EscapeDefaultContext(cfg.Matcher.Params.ScriptName)
  safeJoinToken := shsprintf.EscapeDefaultContext(cfg.Matcher.Params.JoinToken)

  command := fmt.Sprintf("curl -s -L https://%s/v1/webapi/scripts/installer/%s | bash -s %s",
      safeProxyAddr,
      safeScriptName,
      safeJoinToken,
  )
```